### PR TITLE
Improve energy settings dialog

### PR DESCRIPTION
**Pull request by emontnemery**
## Proposed change

Improve energy settings dialogs:
- Remove hard coded units from translation strings and instead fetch units from backend
- Don't include units in input labels
- Fix bugs related to setting up costs for external statistics

Screenshot:
image

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/panels/config/energy/dialogs/dialog-energy-gas-settings.ts [hunk 1]
```diff
@@ -141,14 +160,13 @@ export class DialogEnergyGasSettings
...
           .label=${this.hass.localize(
-            `ui.panel.config.energy.gas.dialog.cost_stat`
+            "ui.panel.config.energy.gas.dialog.cost_stat"
           )}
         >
           <ha-radio
             value="statistic"
             name="costs"
             .checked=${this._costs === "statistic"}
-            .disabled=${externalSource}
```
Comment (emontnemery) [inline src/panels/config/energy/dialogs/dialog-energy-gas-settings.ts:151]: This was wrong: A sensor or external statistics which tracks the total cost is supported both when the consumed gas is tracked by an external statistics or by a sensor. A price entity or a fixed price is only supported when the consumed gas is tracked by a sensor.

Same change for for grid and water.

File: src/panels/config/energy/dialogs/dialog-energy-gas-settings.ts [hunk 2]
```diff
@@ -82,15 +90,19 @@ export class DialogEnergyGasSettings
...
+        : this._gas_units?.join(", ") || "";
+
+    const unitPrice = this._pickedDisplayUnit
+      ? `${this.hass.config.currency}/${this._pickedDisplayUnit}`
+      : undefined;
 
     const externalSource =
-      this._source.stat_cost && isExternalStatistic(this._source.stat_cost);
+      this._source.stat_energy_from &&
+      isExternalStatistic(this._source.stat_energy_from);
```
Comment (emontnemery) [inline src/panels/config/energy/dialogs/dialog-energy-gas-settings.ts:105]: We should check if the gas statistic is external, not if the cost statistic is external.

File: src/panels/config/energy/dialogs/dialog-energy-water-settings.ts [hunk 3]
```diff
@@ -150,35 +187,36 @@ export class DialogEnergyWaterSettings
...
           ></ha-radio>
         </ha-formfield>
         ${this._costs === "number"
           ? html`<ha-textfield
-              .label=${this.hass.localize(
-                `ui.panel.config.energy.water.dialog.cost_number_input`
-              )}
+              .label=${`${this.hass.localize(
+                "ui.panel.config.energy.water.dialog.cost_number_input"
+              )} ${unitPrice ? ` (${unitPrice})` : ""}`}
```
Comment (piitaya) [inline src/panels/config/energy/dialogs/dialog-energy-water-settings.ts:213]:
```suggestion
              .label=${`${this.hass.localize(
                "ui.panel.config.energy.water.dialog.cost_number_input"
              )}${unitPrice ? ` (${unitPrice})` : ""}`}
```

File: src/panels/config/energy/dialogs/dialog-energy-water-settings.ts [hunk 4]
```diff
@@ -150,35 +187,36 @@ export class DialogEnergyWaterSettings
               .hass=${this.hass}
               include-domains='["sensor", "input_number"]'
               .value=${this._source.entity_energy_price}
-              .label=${this.hass.localize(
-                `ui.panel.config.energy.water.dialog.cost_entity_input`
-              )}
+              .label=${`${this.hass.localize(
+                "ui.panel.config.energy.water.dialog.cost_entity_input"
+              )} ${unitPrice ? ` (${unitPrice})` : ""}`}
```
Comment (piitaya) [inline src/panels/config/energy/dialogs/dialog-energy-water-settings.ts:192]:
```suggestion
              .label=${`${this.hass.localize(
                "ui.panel.config.energy.water.dialog.cost_entity_input"
              )}${unitPrice ? ` (${unitPrice})` : ""}`}
```

Comment (piitaya) [reply src/panels/config/energy/dialogs/dialog-energy-water-settings.ts:192]: Only one space is needed

**Review by piitaya — COMMENTED** [review]
I'm ok with the change.
I didn't have the answer for your questions about `external` statistics. May be @bramkragten ?

**Comment by emontnemery** [comment]
> I didn't have the answer for your questions about external statistics. May be @bramkragten ?

I just added the comment to explain the change, the old code was clearly wrong.

**Review by bramkragten — APPROVED** [review]

File: src/panels/config/energy/dialogs/dialog-energy-gas-settings.ts (hunk 1 above)
Comment (bokub) [reply src/panels/config/energy/dialogs/dialog-energy-gas-settings.ts:151]: Hi @emontnemery,

I'm trying to understand why I cannot use a fixed price on external statistics but I cannot find any related documentation.

Could you please tell me where I can find more information about it? I doesn't make sense to me

Thanks!

**Pull request merged**